kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt

```diff
@@ -254,17 +259,23 @@ internal open class ArrayChannel<E>(
     // Note: this function is invoked when channel is already closed
     override fun onCancelIdempotent(wasClosed: Boolean) {
         // clear buffer first, but do not wait for it in helpers
-        if (wasClosed) {
-            lock.withLock {
-                repeat(size.value) {
-                    buffer[head] = 0
-                    head = (head + 1) % buffer.size
+        val onElementCancel = onElementCancel
+        var elementCancelException: ElementCancelException? = null // first cancel exception, others suppressed
+        lock.withLock {
+            repeat(size.value) {
+                val value = buffer[head]
+                if (value !== EMPTY && onElementCancel != null) {
+                    @Suppress("UNCHECKED_CAST")
+                    elementCancelException = onElementCancel.callElementCancelCatchingException(value as E, elementCancelException)
                 }
-                size.value = 0
+                buffer[head] = EMPTY
+                head = (head + 1) % buffer.size
             }
+            size.value = 0
         }
         // then clean all queued senders
         super.onCancelIdempotent(wasClosed)
+        elementCancelException?.let { throw it } // throw cancel exception at the end if there was one
```

Comment:
This behaviour is inconsistent with `cancel` over a regular channel.
Why one rethrows an exception, while the other reports it to global exception handler?
I'd say both should throw an exception from the `cancel` method

Reply:
There's no global exception handler to report the exception in this `cancel`. Channel has no context on its own. In a cancel method, though, the caller up the stack is unreleated to the channel (it is some code cancelling a job), so it makes not sense to force it to handle an exception, but we have a context with an exception handler on our hands.